### ISSUE TITLE
jokr - lastFeeMintTime should update only when _availableManagerFee is greater than 0.

### ISSUE THREAD

**Issue by KupiaSecAdmin** [KupiaSec/sherlock-dhedge-judging-1#87]
jokr

high

# lastFeeMintTime should update only when _availableManagerFee is greater than 0.

## Summary
A vulnerability in the fee calculation allows users to reduce the manager's streaming fee by calling `mintManagerFee` or depositing when `totalSupply` is zero, leading to a loss of rewards for the manager.
## Vulnerability Detail
When deposit or withdrawal actions are triggered, a manager fee can be minted by calling the external function `mintManagerFee`. This function calculates the available manager fee, mints the fee proportionally to the DAO and the manager, and updates `lastFeeMintTime` to the current `block.timestamp`. The available manager fee comprises the streaming fee and the performance fee.

Consider the calculation of the streaming fee in the provided Solidity snippet:

```solidity
uint256 timeChange = block.timestamp.sub(lastFeeMintTime);
uint256 streamingFee = _tokenSupply.mul(timeChange).mul(_managerFeeNumerator).div(_feeDenominator).div(365 days);
```

From this snippet, it is evident that the streaming fee is directly proportional to the time change. However, a vulnerability exists where a user can reduce this streaming fee by calling `mintManagerFee` or making a deposit when `totalSupply` is zero.

Consider the following detailed scenario:

1. **Initial Pool Deployment**:
   - A new Pool is deployed at timestamp `x`.
   - `lastFeeMintTime` is set to `x`.

2. **First Deposit by Alice**:
   - Five days later (at `x + 5 days`), Alice initiates the first deposit.
   - The deposit function triggers `_mintManagerFee`. Since the `totalSupply` is zero, `_availableManagerFee` returns zero:

```solidity
if (_tokenSupply == 0 || _fundValue == 0) return 0;
```

   - Even though no manager fee is minted, `lastFeeMintTime` is updated:

```solidity
if (daoFee > 0) _mint(daoAddress, daoFee);
if (managerFee > 0) _mint(manager(), managerFee);

uint256 currentTokenPrice = _tokenPrice(fundValue, tokenSupply);
if (tokenPriceAtLastFeeMint < currentTokenPrice) {
    tokenPriceAtLastFeeMint = currentTokenPrice;
}

lastFeeMintTime = block.timestamp;
```

   - `lastFeeMintTime` is now `x + 5 days`.

3. **Subsequent Call to `mintManagerFee`**:
   - Suppose someone calls `mintManagerFee` at timestamp `y` (let’s assume `y = x + 10 days`).
   - The manager fee is calculated based on the time change `y - (x + 5 days)` rather than `y - x`.
   - Therefore, the time change used in the calculation is `10 days - 5 days = 5 days` instead of `10 days`.

### Calculation of Loss

Assume the following parameters for simplicity:
- `_tokenSupply = 1000`
- `_managerFeeNumerator = 2`
- `_feeDenominator = 100`

1. **Correct Fee Calculation**:
   - Time change should be `10 days`.
   - Streaming fee: 

```solidity
uint256 correctTimeChange = 10 days;
uint256 correctStreamingFee = _tokenSupply.mul(correctTimeChange).mul(_managerFeeNumerator).div(_feeDenominator).div(365 days);
```

2. **Actual Fee Calculation**:
   - Time change used is `5 days`.
   - Streaming fee:

```solidity
uint256 actualTimeChange = 5 days;
uint256 actualStreamingFee = _tokenSupply.mul(actualTimeChange).mul(_managerFeeNumerator).div(_feeDenominator).div(365 days);
```

3. **Calculation**:

```solidity
uint256 correctStreamingFee = 1000 * 10 days * 2 / 100 / 365 days = 0.5479 (approx)
uint256 actualStreamingFee = 1000 * 5 days * 2 / 100 / 365 days = 0.2739 (approx)
```

### Loss Calculation

The loss in manager rewards is the difference between the correct streaming fee and the actual streaming fee:

```solidity
uint256 loss = correctStreamingFee - actualStreamingFee = 0.5479 - 0.2739 = 0.274 (approx)
```

This example demonstrates how the vulnerability results in a loss of manager rewards. The manager receives 0.274 tokens less than they should have, illustrating the potential impact of this issue over multiple occurrences and larger time frames.
## Impact
Loss of rewards for manager

## Code Snippet
https://github.com/sherlock-audit/2024-05-dhedge/blob/main/V2-Public/contracts/PoolLogic.sol#L723
https://github.com/sherlock-audit/2024-05-dhedge/blob/main/V2-Public/contracts/PoolLogic.sol#L662
 ```solidity
    if (daoFee > 0) _mint(daoAddress, daoFee);

    if (managerFee > 0) _mint(manager(), managerFee);

    uint256 currentTokenPrice = _tokenPrice(fundValue, tokenSupply);
    if (tokenPriceAtLastFeeMint < currentTokenPrice) {
      tokenPriceAtLastFeeMint = currentTokenPrice;
    }

    lastFeeMintTime = block.timestamp;
```

```solidity
    if (lastFeeMintTime != 0) {
      uint256 timeChange = block.timestamp.sub(lastFeeMintTime);
      uint256 streamingFee = _tokenSupply.mul(timeChange).mul(_managerFeeNumerator).div(_feeDenominator).div(365 days);
      available = available.add(streamingFee);
    }
```
## Tool used

Manual Review

## Recommendation
update `lastFeeMintTime` only when `_availbleManagerFee > 0`

**Comment by KupiaSecAdmin** [comment]
@rashtrakoff @D-Ig 

This seems intended and invalid. I agree with what @santipu03 had highlighted to me, that is streaming fee should not collect fees during periods where the Vault doesn't have any deposits, which would mean the manager would be collecting fees for a period where the Vault has no funds, leading to users depositing later paying extra streaming fees.

**Comment by KupiaSecAdmin** [comment]
Yes @nevillehuang I agree. This should be invalid.
